versioning/versioning.rb

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env ruby
```

Comment:
> I find it a bit overkill to wrap this in a Ruby script

Just for giggles, here is the equivalent bash script. It is a bit longer than 10 lines to keep it readable:

```
#!/usr/bin/env bash

set -o errexit -o nounset -o pipefail

function fail {
    echo "$1" >&2
    exit 1
}

if ! type -t git >/dev/null; then
    fail "Cannot find 'git' command"
fi

if ! git rev-parse --is-inside-work-tree &>/dev/null; then
    fail "Directory '${PWD}' is not inside a git work tree"
fi

VERSION=$(git describe --tags --dirty)
if [[ ! "${VERSION}" =~ ^v[0-9]+(\.[0-9]+){2}(-|$) ]]; then
    fail "Tag ${VERSION%-*-g*} is not a semver core version i.e. doesn't match vX.Y.Z"
fi

echo "${VERSION#v}"
```

I realized that we really only support tags of the `v<core version>` format, so the regex checking can be a bit simpler since we don't allow `<pre-release>` and `<build>` elements in the tag names.

Reply:
After switching to `git describe` I also wondered about this too but at the same time it is still just one file and you can have tests and change the functionality later easily if needed.
And of course it was a requirement of the task.

Reply:
> you can have tests

That is the one thing that prevents me from pushing for the simpler bash version. Even the bash version is overkill; as the first 2 conditionals are just re-writing an error condition in a slightly nicer human readable one; the conditions themselves would be caught by `git describe` anyways.

> change the functionality later easily if needed

I have strong feelings (hopes?) that this should not be needed, except for maybe trivial changes.